torch/csrc/jit/frontend/script_type_parser.cpp

```diff
@@ -70,6 +70,30 @@ TypePtr ScriptTypeParser::subscriptToType(
     auto elem_type =
         parseTypeFromExprImpl(*subscript.subscript_exprs().begin());
     return RRefType::create(elem_type);
+  } else if (typeName == "Union") {
+    // In Python 3.9+, Union[NoneType, T] or Union[T, NoneType] are
+    // treated as Optional[T]. Adding the same support for Union in Torchscript.
+    const char* const err =
+        "General Union types are not currently supported."
+        " Only Union[T, NoneType] (i.e. Optional[T]) is "
+        "supported.";
+    if (subscript.subscript_exprs().size() != 2) {
+      throw ErrorReport(subscript) << (err);
+    }
+    auto first_type = parseTypeFromExprImpl(subscript.subscript_exprs()[0]);
+    auto second_type = parseTypeFromExprImpl(subscript.subscript_exprs()[1]);
+
+    bool first_none = first_type == NoneType::get();
+    bool second_none = second_type == NoneType::get();
+
+    if (first_none && !second_none) {
+      return OptionalType::create(second_type);
+    } else if (!first_none && second_none) {
+      return OptionalType::create(first_type);
+    } else {
+      throw ErrorReport(subscript.range()) << err;
```

Comment:
Just to be clear, `Union[None, None]` would trigger this line and error out, is this intended?

Reply:
Yes. That should be handled during the regular Union.